dynamic_programming/abbreviation.cpp

```diff
@@ -0,0 +1,164 @@
+/**
+ * @file
+ * @brief Implementation of Abbrievation
+ * (https://www.hackerrank.com/challenges/abbr/problem)
+ *
+ * @details
+ * Given two strings, a and b, determine if it's possible to make a equal to
+ * b You can perform the following operations on the string a:
+ * 1. Capitalize zero or more of a's lowercase letters.
+ * 2. Delete all of the remaining lowercase letters in a.
+ *
+ * ##Algorithm
+ * The idea is in the problem statement itself: iterate through characters of
+ * string a and b (for character indexes i and j respectively):
+ * 1. If a[i] and b[j] are equal, then move to next position
+ * 2. If a[i] is lowercase of b[j], then explore two possibilities:
+ *  a. Capitalize or
+ *  a. Skip a[i]
+ * 3. If the a[i] is not uppercase, just discard the char, else return false
+ *
+ * Time Complexity (O(|a|*|b|)) where |a| => length of string
+ * @author Ashish Daulatabad (https://github.com/AshishYUO)
+ */
+
+#include <cassert>
+#include <iostream>
+#include <string>
+#include <vector>
+/**
+ * @namespace dynamic_programming
+ * @brief Dynamic Programming Algorithms
+ */
+namespace dynamic_programming {
+/**
+ * @namespace Knapsack
+ * @brief Implementation of Abbreivation problem
+ */
+namespace abbreviation {
+/**
+ *  (recursive dp function) https://www.hackerrank.com/challenges/abbr/problem
+ *  Returns whether s can be converted to t with following rules:
+ *  a. Capitalize zero or more of a's lowercase letters from string s
+ *  b. remove all other lowercase letters from string s
+ * @param dp: memo as parameter to store the result
+ * @param v: visited boolean to check if the result is already computed
+ * @param s: given string
+ * @param t: resultant abbreivated string
+ * @param i: start of string s
+ * @param j: start of string j
+ * @returns bool whether s can be converted to t
+ */
+bool abbreviation_recursion(std::vector<std::vector<bool>> *memo,
+                            std::vector<std::vector<bool>> *visited,
+                            const std::string &s, const std::string &t,
+                            int i = 0, int j = 0) {
```

Comment:
I don't think these should be parameters. They should be variables inside the function.

Reply:
I don't know how to do that actually.

If the intention is to not let user pass memo and visited and let function handle it, then I can create a helper function for that.

Reply:
Just create two variables inside the function.
```suggestion
) {
    int i = 0, j = 0;
```

Reply:
These two variables are helpful for resolving the base case or changing the state from one to another.
E.g., lines 74, 76, and 77.

I kept them default to zero if the user does not pass them explicitly.

Reply:
Yes, that's good, and it looks fine for me. But why not use variables instead of parameters?

Reply:
I don't know if there are other ways around to call recursion without `i` and `j` as parameters.
I can only think of them by passing them as parameters.